### PR TITLE
Mypy fix/io sql

### PR DESCRIPTION
**Pull request by aaditya-panik**
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

As part of this PR, did the following:

* Variable typing.
* Code cleanup and reformatting the imports.
* Internal class `_TestSQLAlchemyConn` behaves like an abstract class.
* To solve **variable name clashing** between parent classes in some of the errors, I restructured the inheritance from a structure like 

  > (_TestSQLAlchemy, _TestDBSQLAlchemy) --> TestDBSQLAlchemy

  to

  > _TestSQLAlchemy --> _TestDBSQLAlchemy --> TestDBSQLAlchemy

  Not sure why it was like that unless I'm missing something?

### CODE REVIEW

File: pandas/tests/io/test_sql.py [hunk 1]
```diff
@@ -582,8 +584,8 @@ class _TestSQLApi(PandasSQLTest):
 
     """
 
-    flavor = "sqlite"
-    mode = None
+    flavor = "sqlite"  # type: Optional[str]
+    mode = None  # type: Optional[str]
```
Comment (simonjayhawkins) [inline pandas/tests/io/test_sql.py]:
```suggestion
    mode = None  # type: str
```

File: pandas/tests/io/test_sql.py [hunk 2]
```diff
@@ -1234,7 +1236,7 @@ class _TestSQLAlchemy(SQLAlchemyMixIn, PandasSQLTest):
 
     """
 
-    flavor = None
+    flavor = None  # type: Optional[str]
```
Comment (simonjayhawkins) [inline pandas/tests/io/test_sql.py]:
```suggestion
    flavor = None  # type: str
```

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
@aaditya-panik Thanks for the PR. 

I _think_ is you make the changes suggested, all the other changes can be reverted.

**Comment by aaditya-panik** [comment]
I will be creating and linking a new PR because I think I botched my forked repo branch. Will link the two PRs once I create the new one with the changes @simonjayhawkins suggested.

**Comment by aaditya-panik** [comment]
@simonjayhawkins made the changes you suggested. Let me know if that works.

**Pull request closed without merging**